tests/instruments/test_all_instruments.py

```diff
@@ -23,53 +23,58 @@
 #
 
 import importlib
+from pathlib import Path
 import pytest
 from unittest.mock import MagicMock
 
 from pymeasure import instruments
 from pymeasure.instruments import Instrument, Channel
 
-
 # Collect all instruments
-devices = []
-channels = []
+devices = set()
+channels = set()
 
 
 def find_devices_in_module(module, devices, channels):
-    for dev in dir(module):
-        if dev.startswith("__"):
-            continue
-        d = getattr(module, dev)
+    base_dir = Path(module.__path__[0])
+    base_import = module.__package__ + '.'
+    for folder in [i for i in base_dir.iterdir() if i.is_dir() and i.name[0] != '.']:
         try:
-            i = issubclass(d, Instrument)
-            c = issubclass(d, Channel)
-        except TypeError:
-            # d is no class
-            continue
-        else:
-            if i and d not in devices:
-                devices.append(d)
-            elif c and d not in channels:
-                channels.append(d)
+            package = importlib.import_module(base_import + folder.name).__package__ + '.'
+            for mod in [f for f in folder.iterdir() if
+                        f.name[-3:] == '.py' and f.name != '__init__.py']:
+                mod_name = package + mod.name[:-3]
+                submodule = importlib.import_module(mod_name)
+                for dev in dir(submodule):
+                    if dev.startswith("__"):
+                        continue
+                    d = getattr(submodule, dev)
+                    try:
+                        i = issubclass(d, Instrument)
+                        c = issubclass(d, Channel)
+                    except TypeError:
+                        # d is no class
+                        continue
+                    else:
+                        if i:
+                            devices.add(d)
+                        elif c:
+                            channels.add(d)
+        except ModuleNotFoundError:
+            # Some non-required driver dependencies may not be installed on test computer,
+            # for example ni.VirtualBench
+            pass
+        except OSError:
+            # On Windows instruments.ni.daqmx can raise an OSError before ModuleNotFoundError
+            # when checking installed driver files
+            pass
 
 
 find_devices_in_module(instruments, devices, channels)  # the instruments module itself
-for manufacturer in dir(instruments):
-    if manufacturer.startswith("__"):
-        continue
-    manu = getattr(instruments, manufacturer)
-    find_devices_in_module(manu, devices, channels)  # module in instruments package
```

Comment:
Do you look for devices the `instruments` package itself, for example `Instrument` in `instrument.py`, `CommonBase`, `Channel` etc.?
If I remember correctly, I encountered 123 instruments and some 20 channels. (you can print that list in a test which fails 😁 )

For that reason, I ran the routine over instruments imported directly into `instruments`, over all modules in `instruments` and over all modules in all packages in `instruments`.

If you use that finding algorithm, maybe a recursive approach is best, in order to find all instruments and channels?

Reply:
Yea, when writing this I was comparing the numbers and got the same amount of 123 instruments and 20 channels. `Instrument` and `Channel` are imported through the submodule files.

It did seem like a good use for recursion while writing it. I just pushed a change to use `Path.rglob`, which itself returns a recursive list of files and folders based on a pattern. Now instead of making `find_devices_in_module` recursive, it just iterates over the python files that `Path.rglob` returns.

Now 125 instruments and 20 channels are returned, this method caught `pymeasure.instruments.fakes.SwissArmyFake` and `pymeasure.instruments.fakes.FakeInstrument`, I hadn't been looking at `fakes.py` before. I reckon the fake instruments should still follow all docstring styles and such, so it can be useful to include them.

Reply:
The docstring tests for `pymeasure.instruments.fakes.SwissArmyFake` and `pymeasure.instruments.fakes.FakeInstrument` did in fact fail which is pretty funny. Updated their docstrings and added them to `init_communication` for their fake adapters.